pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -1052,21 +1154,90 @@ func (configgen *ConfigGeneratorImpl) buildSidecarOutboundListenerForPortOrUDS(l
 	var ret bool
 	var opts []*filterChainOpts
 
+	conflictType := NoConflict
+
 	switch pluginParams.ListenerProtocol {
 	case plugin.ListenerProtocolHTTP:
-		if ret, opts = configgen.buildSidecarOutboundHTTPListenerOptsForPortOrUDS(&listenerMapKey, &currentListenerEntry,
+		if ret, opts = configgen.buildSidecarOutboundHTTPListenerOptsForPortOrUDS(node, &listenerMapKey, &currentListenerEntry,
 			&listenerOpts, pluginParams, listenerMap, actualWildcard); !ret {
 			return
 		}
 
+		// Check if conflict happens
+		if util.IsProtocolSniffingEnabled(node) && currentListenerEntry != nil {
+			// Build HTTP listener. If current listener entry is using HTTP or protocol sniffing,
+			// append the service. Otherwise (TCP), change current listener to use protocol sniffing.
+			if currentListenerEntry.protocol.IsHTTP() {
+				conflictType = HTTPOverHTTP
+			} else if currentListenerEntry.protocol.IsTCP() {
+				conflictType = HTTPOverTCP
+			} else {
+				conflictType = HTTPOverAuto
+			}
+		}
+
 		listenerOpts.filterChainOpts = opts
+
 	case plugin.ListenerProtocolTCP:
-		if ret, opts = configgen.buildSidecarOutboundTCPListenerOptsForPortOrUDS(&destinationCIDR, &listenerMapKey, &currentListenerEntry,
+		if ret, opts = configgen.buildSidecarOutboundTCPListenerOptsForPortOrUDS(node, &destinationCIDR, &listenerMapKey, &currentListenerEntry,
 			&listenerOpts, pluginParams, listenerMap, virtualServices, actualWildcard); !ret {
 			return
 		}
 
+		// Check if conflict happens
+		if util.IsProtocolSniffingEnabled(node) && currentListenerEntry != nil {
+			// Build TCP listener. If current listener entry is using HTTP, add a new TCP filter chain
+			// If current listener is using protocol sniffing, merge the TCP filter chains.
+			if currentListenerEntry.protocol.IsHTTP() {
+				conflictType = TCPOverHTTP
+			} else if currentListenerEntry.protocol.IsTCP() {
+				conflictType = TCPOverTCP
+			} else {
+				conflictType = TCPOverAuto
+			}
+		}
+
 		listenerOpts.filterChainOpts = opts
+
+	case plugin.ListenerProtocolAuto:
+		// Add tcp filter chain
+		if ret, opts = configgen.buildSidecarOutboundTCPListenerOptsForPortOrUDS(node, &destinationCIDR, &listenerMapKey, &currentListenerEntry,
+			&listenerOpts, pluginParams, listenerMap, virtualServices, actualWildcard); !ret {
+			return
+		}
+		listenerOpts.filterChainOpts = append(listenerOpts.filterChainOpts, opts...)
+
+		// Add http filter chain and tcp filter chain to the listener opts
+		// Build HTTP filter chain first. If there is already another listener
```

Comment:
you are building tcp filter chain first but comment states otherwise

Reply:
Good catch! The HTTP filter chain was put before TCP filter chain at the very beginning. I didn't check the comment. Thanks!